File: phone-append-2-nodejs/REST/composite_phone_append_rest.js
import axios from 'axios';
import querystring from 'querystring';
import { PA2Response } from './pa2_response.js';

/**
 * @constant
 * @type {string}
 * @description The base URL for the live ServiceObjects PhoneAppend2 (PA2) API service.
 */
const LiveBaseUrl = 'https://sws.serviceobjects.com/pa2/api.svc/json/';

/**
 * @constant
 * @type {string}
 * @description The base URL for the backup ServiceObjects PhoneAppend2 (PA2) API service.
 */
const BackupBaseUrl = 'https://swsbackup.serviceobjects.com/pa2/api.svc/json/';

/**
 * @constant
 * @type {string}
 * @description The base URL for the trial ServiceObjects PhoneAppend2 (PA2) API service.
 */
const TrialBaseUrl = 'https://trial.serviceobjects.com/pa2/api.svc/json/';

/**
 * <summary>
 * Checks if a response from the API is valid by verifying that it either has no Error object
 * or the Error.TypeCode is not equal to '3'.
 * </summary>
 * <param name="response" type="Object">The API response object to validate.</param>
 * <returns type="boolean">True if the response is valid, false otherwise.</returns>
 */
const isValid = (response) => !response?.Error || response.Error.TypeCode !== '3';

/**
 * <summary>
 * Constructs a full URL for the CompositePhoneAppendJson API endpoint by combining the base URL
 * with query parameters derived from the input parameters.
 * </summary>
 * <param name="params" type="Object">An object containing all the input parameters.</param>
 * <param name="baseUrl" type="string">The base URL for the API service (live, backup, or trial).</param>
 * <returns type="string">The constructed URL with query parameters.</returns>
 */
const buildUrl = (params, baseUrl) =>
    `${baseUrl}CompositePhoneAppendJson?${querystring.stringify(params)}`;

/**
 * <summary>
 * Performs an HTTP GET request to the specified URL with a given timeout.
 * </summary>
 * <param name="url" type="string">The URL to send the GET request to.</param>
 * <param name="timeoutSeconds" type="number">The timeout duration in seconds for the request.</param>
 * <returns type="Promise<PA2Response>">A promise that resolves to a PA2Response object containing the API response data.</returns>
 * <exception cref="Error">Thrown if the HTTP request fails, with a message detailing the error.</exception>
 */
const httpGet = async (url, timeoutSeconds) => {
    try {
        const response = await axios.get(url, { timeout: timeoutSeconds * 1000 });
        return new PA2Response(response.data);
    } catch (error) {
        throw new Error(`HTTP request failed: ${error.message}`);
    }
};

/**
 * <summary>
 * Provides functionality to call the ServiceObjects PhoneAppend2 (PA2) API's CompositePhoneAppendJson endpoint,
 * retrieving phone number information for a contact (business or residential) based on provided inputs with fallback to a backup endpoint for reliability in live mode.
 * </summary>
 */
const CompositePhoneAppendClient = {
    /**
     * <summary>
     * Asynchronously invokes the CompositePhoneAppendJson API endpoint, attempting the primary endpoint
     * first and falling back to the backup if the response is invalid (Error.TypeCode == '3') in live mode.
     * </summary>
     * @param {string} Name - The name of the contact.
     * @param {string} Address - Address line of the contact. Optional.
     * @param {string} City - The city of the contact.
     * @param {string} State - The state of the contact
     * @param {string} PostalCode - The postal code of the contact (e.g., "93101"). Optional.
     * @param {string} IsBusiness - Indicates if the contact is a business ("True", "False", or empty). Optional.
     * @param {string} LicenseKey - Your license key to use the service.
     * @param {boolean} isLive - Value to determine whether to use the live or trial servers.
     * @param {number} timeoutSeconds - Timeout, in seconds, for the call to the service.
     * @returns {Promise<PA2Response>} - A promise that resolves to a PA2Response object.
     */
    async invokeAsync(Name, Address, City, State, PostalCode, IsBusiness, LicenseKey, isLive = true, timeoutSeconds = 15) {
        const params = {
            Name,
            Address,
            City,
            State,
            PostalCode,
            IsBusiness,
            LicenseKey
        };

        const url = buildUrl(params, isLive ? LiveBaseUrl : TrialBaseUrl);
        let response = await httpGet(url, timeoutSeconds);

        if (isLive && !isValid(response)) {
            const fallbackUrl = buildUrl(params, BackupBaseUrl);
            const fallbackResponse = await httpGet(fallbackUrl, timeoutSeconds);
            return fallbackResponse;
        }
        return response;
    },

    /**
     * <summary>
     * Synchronously invokes the CompositePhoneAppendJson API endpoint by wrapping the async call
     * and awaiting its result immediately.
     * </summary>
     * @param {string} Name - The name of the contact.
     * @param {string} Address - Address line of the contact. Optional.
     * @param {string} City - The city of the contact.
     * @param {string} State - The state of the contact.
     * @param {string} PostalCode - The postal code of the contact. Optional.
     * @param {string} IsBusiness - Indicates if the contact is a business ("True", "False", or empty). Optional.
     * @param {string} LicenseKey - Your license key to use the service.
     * @param {boolean} isLive - Value to determine whether to use the live or trial servers.
     * @param {number} timeoutSeconds - Timeout, in seconds, for the call to the service.
     * @returns {PA2Response} - A PA2Response object with phone number details or an error.
     */
    invoke(Name, Address, City, State, PostalCode, IsBusiness, LicenseKey, isLive = true, timeoutSeconds = 15) {
        return (async () => await this.invokeAsync(
            Name, Address, City, State, PostalCode, IsBusiness, LicenseKey, isLive, timeoutSeconds
        ))();
    }
};

export { CompositePhoneAppendClient, PA2Response };

File: phone-append-2-python/REST/composite_phone_append_rest.py
from pa2_response import PA2Response, PhoneInfo, Error
import requests
from typing import Optional

# Endpoint URLs for ServiceObjects CompositePhoneAppend API
primary_url = "https://sws.serviceobjects.com/pa2/api.svc/json/CompositePhoneAppendJson?"
backup_url = "https://swsbackup.serviceobjects.com/pa2/api.svc/json/CompositePhoneAppendJson?"
trial_url = "https://trial.serviceobjects.com/pa2/api.svc/json/CompositePhoneAppendJson?"

def get_composite_phone_append(
    name: str,
    address: Optional[str] = None,
    city: Optional[str] = None,
    state: Optional[str] = None,
    postal_code: Optional[str] = None,
    is_business: Optional[str] = None,
    license_key: Optional[str] = None,
    is_live: bool = True
) -> PA2Response:
    """
    Call ServiceObjects CompositePhoneAppend API to retrieve a phone number for a given contact (business or residential).

    Parameters:
        name: The name of the contact (business or individual) to retrieve the phone number.
        address: Address line of the contact. Optional.
        city: The city of the contact. Optional.
        state: The state of the contact. Optional.
        postal_code: The postal code of the contact. Optional.
        is_business: Specifies if the contact is a business ("True", "False", or None to let the service decide). Optional.
        license_key: Your ServiceObjects license key.
        is_live: Use live or trial servers.

    Returns:
        PA2Response: Parsed JSON response with phone information or error details.

    Raises:
        RuntimeError: If the API returns an error payload.
        requests.RequestException: On network/HTTP failures (trial mode).
    """
    params = {
        "Name": name,
        "Address": address,
        "City": city,
        "State": state,
        "PostalCode": postal_code,
        "IsBusiness": is_business,
        "LicenseKey": license_key,
    }
    # Select the base URL: production vs trial
    url = primary_url if is_live else trial_url

    try:
        # Attempt primary (or trial) endpoint
        response = requests.get(url, params=params, timeout=10)
        response.raise_for_status()
        data = response.json()

        # If API returned an error in JSON payload, trigger fallback
        error = data.get('Error')
        if not (error is None or error.get('TypeCode') != "3"):
            if is_live:
                # Try backup URL
                response = requests.get(backup_url, params=params, timeout=10)
                response.raise_for_status()
                data = response.json()

                # If still error, propagate exception
                if 'Error' in data:
                    raise RuntimeError(f"CompositePhoneAppend service error: {data['Error']}")
            else:
                # Trial mode error is terminal
                raise RuntimeError(f"CompositePhoneAppend trial error: {data['Error']}")

        # Convert JSON response to PA2Response for structured access
        error = Error(**data.get("Error", {})) if data.get("Error") else None

        return PA2Response(
            PhoneInfo=PhoneInfo(
                Phone=data.get("PhoneInfo", {}).get("Phone"),
                Name=data.get("PhoneInfo", {}).get("Name"),
                Address=data.get("PhoneInfo", {}).get("Address"),
                City=data.get("PhoneInfo", {}).get("City"),
                State=data.get("PhoneInfo", {}).get("State"),
                PostalCode=data.get("PhoneInfo", {}).get("PostalCode"),
                IsResidential=data.get("PhoneInfo", {}).get("IsResidential"),
                Certainty=data.get("PhoneInfo", {}).get("Certainty"),
                LineType=data.get("PhoneInfo", {}).get("LineType"),
                Debug=data.get("PhoneInfo", {}).get("Debug")
            ) if data.get("PhoneInfo") else None,
            Error=error
        )

    except requests.RequestException as req_exc:
        # Network or HTTP-level error occurred
        if is_live:
            try:
                # Fallback to backup URL
                response = requests.get(backup_url, params=params, timeout=10)
                response.raise_for_status()
                data = response.json()
                if "Error" in data:
                    raise RuntimeError(f"CompositePhoneAppend backup error: {data['Error']}") from req_exc

                error = Error(**data.get("Error", {})) if data.get("Error") else None

                return PA2Response(
                    PhoneInfo=PhoneInfo(
                        Phone=data.get("PhoneInfo", {}).get("Phone"),
                        Name=data.get("PhoneInfo", {}).get("Name"),
                        Address=data.get("PhoneInfo", {}).get("Address"),
                        City=data.get("PhoneInfo", {}).get("City"),
                        State=data.get("PhoneInfo", {}).get("State"),
                        PostalCode=data.get("PhoneInfo", {}).get("PostalCode"),
                        IsResidential=data.get("PhoneInfo", {}).get("IsResidential"),
                        Certainty=data.get("PhoneInfo", {}).get("Certainty"),
                        LineType=data.get("PhoneInfo", {}).get("LineType"),
                        Debug=data.get("PhoneInfo", {}).get("Debug")
                    ) if data.get("PhoneInfo") else None,
                    Error=error
                )
            except Exception as backup_exc:
                raise RuntimeError("CompositePhoneAppend service unreachable on both endpoints") from backup_exc
        else:
            raise RuntimeError(f"CompositePhoneAppend trial error: {str(req_exc)}") from req_exc